File: python/particle.py
import random

class Particle:
    """Particle to be simulated

    The class stores postions and velocities in x and y
    outside of a list for better performance.

    Atttibutes:
        posX: A float indicating the x position of the particle
        posY: A float indicating the y position of the particle
        velX: A float indicating the x velocity of the particle
        velY: A float indicating the y velocity of the particle
        radius: A float indicating the fixed radius of the particle
    """

    radius = 0.1
    def __init__(self, pos, vel=None):
        """Initializes the instance with an initial position and velocity.

        vel is initialized as None so that each particle can be assigned
        a different random velocity by default.

        Args:
            pos: List with initial position
            vel: List with initial velocity
        """
        self.posX = pos[0]
        self.posY = pos[1]

        if vel is None:
            vel = [random.random() for _ in range(2)]

        self.velX = vel[0]
        self.velY = vel[1]

    def move(self, dt):
        self.posX += dt*self.velX
        self.posY += dt*self.velY

    def edgeColision(self):
        """Checks for colision with edges."""
        if self.posX <= self.radius:
            self.velX = abs(self.velX)
        elif self.posX >= 1-self.radius:
            self.velX = -abs(self.velX)

        if self.posY <= self.radius:
            self.velY = abs(self.velY)
        elif self.posY >= 1-self.radius:
            self.velY = -abs(self.velY)

File: c++/particle.cpp
#include <random>
#include <array>
#include <cmath>

#include "particle.hpp"

Particle::Particle(){
  radius = 0.05;

  // Generates a non-deterministic number between 0.2 and 1
  // as the initial velocity
  std::random_device rd;
  std::mt19937 gen(rd());
  std::uniform_real_distribution<> dis(0.1, 0.9);

  pos = {dis(gen), dis(gen), dis(gen)};
  vel = {dis(gen), dis(gen), dis(gen)};
}

Particle::Particle(double R){
  radius = R;

  // Generates a non-deterministic number between 0.2 and 1
  // as the initial velocity
  std::random_device rd;
  std::mt19937 gen(rd());
  std::uniform_real_distribution<> dis(0.1, 0.9);

  pos = {dis(gen), dis(gen), dis(gen)};
  vel = {dis(gen), dis(gen), dis(gen)};
}

Particle::Particle(std::array<double,3> position){
  radius = 0.05;
  pos = position;

  // Generates a non-deterministic number between 0.2 and 1
  // as the initial velocity
  std::random_device rd;
  std::mt19937 gen(rd());
  std::uniform_real_distribution<> dis(0.2, 1.0);

  vel = {dis(gen), dis(gen), dis(gen)};
}

Particle::Particle(std::array<double,3> position, std::array<double,3> velocity){
  radius = 0.05;
  pos = position;
  vel = velocity;
}

void Particle::move(double dt){
  for (int i=0; i<3; i++){
    pos[i] += dt*vel[i];
  }
}

void Particle::edgeColision(){
  for (int i=0; i<3; i++){
    if (pos[i] <= radius){
      vel[i] = std::abs(vel[i]);
    }
    else if (pos[i] >= 1-radius){
      vel[i] = -std::abs(vel[i]);
    }
  }
}


